torchrl/collectors/distributed/ray_collector.py

```diff
@@ -0,0 +1,321 @@
+            ray_init_config = DEFAULT_RAY_INIT_CONFIG
+
+        if not _has_ray:
+            raise RuntimeError(
+                f"ray library not found, unable to create a DistributedCollector. "
+            ) from RAY_ERR
+
+        if coordination not in ("sync", "async"):
+            raise ValueError(f"Coordination input parameter in CollectorSet has to be sync or async.")
+
+        # Connect to an existing Ray cluster or start one and connect to it.
+        ray.init(**ray_init_config)
+        if not ray.is_initialized():
+            raise RuntimeError(f"Ray could not be initialized.")
+
+        # Monkey patching as_remote and print_remote_collector_info to collector class
+        collector_class.as_remote = as_remote
+        collector_class.print_remote_collector_info = print_remote_collector_info
+        self.collector_class = collector_class
+
+        self.collected_frames = 0
+        self.total_frames = total_frames
+        self.collector_class = collector_class
+        self.collector_params = collector_params
+        self.num_collectors = num_collectors
+        self.remote_config = remote_config
+        self.coordination = coordination
+        self.local_policy = policy
+
+        # Create remote instances of the collector class
+        self._remote_collectors = []
+        if self.num_collectors > 0:
+            self.add_collectors(self.num_collectors, collector_params)
+
+        # Print info of all remote workers
+        pending_samples = [e.print_remote_collector_info.remote() for e in self.remote_collectors()]
+        ray.wait(object_refs=pending_samples)
+
+    @staticmethod
+    def _make_collector(cls, collector_params):
+        """Create a single collector instance."""
+        collector = cls(**collector_params)
+        return collector
+
+    def add_collectors(self, num_collectors, collector_params):
+        """Creates and adds a number of remote collectors to the set."""
+        cls = self.collector_class.as_remote(**self.remote_config).remote
+        self._remote_collectors.extend(
+            [self._make_collector(cls, collector_params) for _ in range(num_collectors)])
+
+    def local_policy(self):
+        """Returns local collector."""
+        return self.local_policy
+
+    def remote_collectors(self):
+        """Returns list of remote collectors."""
+        return self._remote_collectors
+
+    def stop_remote_collectors(self):
+        """Stops all remote collectors."""
+        for collector in self.remote_collectors():
+            # collector.__ray_terminate__.remote()  # This will kill the actor but let pending tasks finish
+            ray.kill(collector)  # This will interrupt any running tasks on the actor, causing them to fail immediately
+
+    def iterator(self):
+        if self.coordination == "sync":
+            return self._sync_iterator()
+        else:
+            return self._async_iterator()
+
+    def _sync_iterator(self) -> Iterator[TensorDictBase]:
```

Comment:
where do we update the weights here?

Reply:
here:
https://github.com/pytorch/rl/blob/b3a8ec71eeeaf00998ca68ddb9216063a1dd2a1a/torchrl/collectors/distributed/ray_collector.py#L209